File: Samples/MapScenariosObjC/MapScenariosObjC/Scenarios/LocationModesViewController.m
//
//  LocationModesViewController.m
//  MapScenariosObjC
//
//  Created on 3/6/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

#import <PWMapKit/PWMapKit.h>
#import <PWCore/PWCore.h>

#import "LocationModesViewController.h"

@implementation UIImage (TrackingMode)

+ (UIImage *)emptyTrackingImage:(UIColor *)color {
    CGRect rect = CGRectMake(0, 0, 26, 26);
    
    UIGraphicsBeginImageContextWithOptions(rect.size, NO, 0);
    
    UIBezierPath *bezierPath = [UIBezierPath new];
    [bezierPath moveToPoint:CGPointMake(13.53, 24)];
    [bezierPath addLineToPoint:CGPointMake(13.44, 10.62)];
    [bezierPath addLineToPoint:CGPointMake(0, 10.71)];
    [bezierPath addLineToPoint:CGPointMake(24, 0)];
    [bezierPath addLineToPoint:CGPointMake(13.53, 24)];
    [bezierPath closePath];
    [bezierPath moveToPoint:CGPointMake(14.31, 9.74)];
    [bezierPath addLineToPoint:CGPointMake(14.38, 19.85)];
    [bezierPath addLineToPoint:CGPointMake(22.29, 1.7)];
    [bezierPath addLineToPoint:CGPointMake(4.16, 9.81)];
    [bezierPath addLineToPoint:CGPointMake(14.31, 9.74)];
    [bezierPath closePath];
    bezierPath.miterLimit = 4;
    
    [color setFill];
    [bezierPath fill];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

+ (UIImage *)filledTrackingImage:(UIColor *)color {
    CGRect rect = CGRectMake(0, 0, 26, 26);
    
    UIGraphicsBeginImageContextWithOptions(rect.size, NO, 0);
    
    UIBezierPath *bezierPath = [UIBezierPath new];
    [bezierPath moveToPoint:CGPointMake(12.4, 22)];
    [bezierPath addLineToPoint:CGPointMake(12.33, 9.74)];
    [bezierPath addLineToPoint:CGPointMake(0, 9.81)];
    [bezierPath addLineToPoint:CGPointMake(22, 0)];
    [bezierPath addLineToPoint:CGPointMake(12.4, 22)];
    [bezierPath closePath];
    bezierPath.miterLimit = 4;
    
    [color setFill];
    [bezierPath fill];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

+ (UIImage *)trackWithHeadingImage:(UIColor *)color {
    CGRect rect = CGRectMake(0, 0, 26, 26);
    
    UIGraphicsBeginImageContextWithOptions(rect.size, NO, 0);
    
    UIBezierPath *bezierPath = [UIBezierPath new];
    [bezierPath moveToPoint:CGPointMake(6.98, 0)];
    [bezierPath addLineToPoint:CGPointMake(8, 0)];
    [bezierPath addLineToPoint:CGPointMake(8, 6.01)];
    [bezierPath addLineToPoint:CGPointMake(6.98, 6.01)];
    [bezierPath addLineToPoint:CGPointMake(6.98, 0)];
    [bezierPath closePath];
    [bezierPath moveToPoint:CGPointMake(7.44, 19.01)];
    [bezierPath addLineToPoint:CGPointMake(0, 26)];
    [bezierPath addLineToPoint:CGPointMake(7.39, 8.01)];
    [bezierPath addLineToPoint:CGPointMake(15, 25.91)];
    [bezierPath addLineToPoint:CGPointMake(7.44, 19.01)];
    [bezierPath closePath];
    bezierPath.miterLimit = 4;
    
    [color setFill];
    [bezierPath fill];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

@interface LocationModesViewController () <PWMapViewDelegate>

@property (nonatomic, strong) NSString *applicationId;
@property (nonatomic, strong) NSString *accessKey;
@property (nonatomic, strong) NSString *signatureKey;

@property (nonatomic, strong) PWMapView *mapView;
@property (weak, nonatomic) IBOutlet UIToolbar *toolbar;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *trackingModeButton;

@end

@implementation LocationModesViewController

- (id)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        _buildingIdentifier = 0; // Enter your building identifier here, found on the building's Edit page on Maas portal
        // Enter your application identifier, access key, and signature key, found on Maas portal under Account > Apps
        _applicationId = @"";
        _accessKey = @"";
        _signatureKey = @"";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Location Modes";
    
    if (self.applicationId.length > 0 && self.accessKey.length > 0 && self.signatureKey.length > 0) {
        [PWCore setApplicationID:self.applicationId accessKey:self.accessKey signatureKey:self.signatureKey];
    }
    
    self.mapView = [PWMapView new];
    self.mapView.delegate = self;
    [self.view addSubview:self.mapView];
    [self configureMapViewConstraints];
    
    self.trackingModeButton.image = [UIImage emptyTrackingImage:[UIColor blueColor]];
    
    [PWBuilding buildingWithIdentifier:self.buildingIdentifier completion:^(PWBuilding *building, NSError *error) {
        __weak typeof(self) weakSelf = self;
        [weakSelf.mapView setBuilding:building animated:YES onCompletion:^(NSError *error) {
            PWManagedLocationManager *managedLocationManager = [[PWManagedLocationManager alloc] initWithBuildingId:weakSelf.buildingIdentifier];
            dispatch_async(dispatch_get_main_queue(), ^{
                [weakSelf.mapView registerLocationManager:managedLocationManager];
            });
        }];
    }];
}

- (void)configureMapViewConstraints {
    self.mapView.translatesAutoresizingMaskIntoConstraints = NO;
    [[self.mapView.topAnchor constraintEqualToAnchor:self.view.topAnchor] setActive:YES];
    [[self.mapView.bottomAnchor constraintEqualToAnchor:self.toolbar.topAnchor] setActive:YES];
    [[self.mapView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor] setActive:YES];
    [[self.mapView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor] setActive:YES];
}

- (IBAction)trackingModeButtonTapped:(id)sender {
    switch (self.mapView.trackingMode) {
        case PWTrackingModeNone:
            self.mapView.trackingMode = PWTrackingModeFollow;
            break;
        case PWTrackingModeFollow:
            self.mapView.trackingMode = PWTrackingModeFollowWithHeading;
            break;
        case PWTrackingModeFollowWithHeading:
            self.mapView.trackingMode = PWTrackingModeNone;
            break;
        default:
            break;
    }
}

#pragma mark - PWMapViewDelegate

- (void)mapView:(PWMapView *)mapView didChangeIndoorUserTrackingMode:(PWTrackingMode)mode {
    switch (mode) {
        case PWTrackingModeNone:
            self.trackingModeButton.image = [UIImage emptyTrackingImage:[UIColor blueColor]];
            break;
        case PWTrackingModeFollow:
            self.trackingModeButton.image = [UIImage filledTrackingImage:[UIColor blueColor]];
            break;
        case PWTrackingModeFollowWithHeading:
            self.trackingModeButton.image = [UIImage trackWithHeadingImage:[UIColor blueColor]];
            break;
        default:
            break;
    }
}

@end

File: Samples/MapScenarios/MapScenarios/Scenarios/LocationModesViewController.swift
//
//  LocationModesViewController.swift
//  MapScenarios
//
//  Created on 3/5/18.
//  Copyright © 2018 Phunware. All rights reserved.
//

import Foundation
import UIKit
import PWCore
import PWLocation
import PWMapKit

// MARK: - LocationModesViewController
class LocationModesViewController: UIViewController, ScenarioProtocol {
    
    // Enter your application identifier and access key found on Maas portal under Account > Apps
    var applicationId = ""
    var accessKey = ""
    
    // Enter your campus identifier here, found on the campus's Edit page on Maas portal
    var campusIdentifier = 0

    // Enter your building identifier here, found on the building's Edit page on Maas portal
    var buildingIdentifier = 0
    
    private let mapView = PWMapView()
    private let locationManager = CLLocationManager()
    
    @IBOutlet weak var toolbar: UIToolbar!
    @IBOutlet weak var trackingModeButton: UIBarButtonItem!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        navigationItem.title = "Location Modes"
        
        if !validateScenarioSettings() {
            return
        }
        
        PWCore.setApplicationID(applicationId, accessKey: accessKey)
        mapView.delegate = self
        view.addSubview(mapView)
        configureMapViewConstraints()
        
        trackingModeButton.image = .emptyTrackingImage(color: .blue)

        // If we want to route between buildings on a campus, then we use PWCampus.campus to configure MapView
        // Otherwise, we will use PWBuilding.building to route between floors in a single building.
        if campusIdentifier != 0 {
            PWCampus.campus(identifier: campusIdentifier) { [weak self] result in
                switch result {
                case .success(let campus):
                    self?.mapView.setCampus(campus, animated: true, onCompletion: { (error) in
                        self?.locationManager.delegate = self
                        if !CLLocationManager.isAuthorized() {
                            self?.locationManager.requestWhenInUseAuthorization()
                        } else {
                            self?.startManagedLocationManager()
                        }
                    })
                    
                case .failure(let error):
                    self?.warning(error.localizedDescription)
                }
            }
        }
        else {
            PWBuilding.building(identifier: buildingIdentifier) { [weak self] result in
                switch result {
                case .success(let building):
                    self?.mapView.setBuilding(building, animated: true, onCompletion: { (error) in
                        self?.locationManager.delegate = self
                        if !CLLocationManager.isAuthorized() {
                            self?.locationManager.requestWhenInUseAuthorization()
                        } else {
                            self?.startManagedLocationManager()
                        }
                    })
                    
                case .failure(let error):
                    self?.warning(error.localizedDescription)
                }
            }
        }
    }
    
    func startManagedLocationManager() {
        
        // In order to route between buildings on a campus, we also need to register the
        // PWManagedLocationManager using campusIdentifier.  Otherwise, we will register
        // using buildingIdentifier.
        if campusIdentifier != 0 {
            DispatchQueue.main.async { [weak self] in
                guard let campusIdentifier = self?.campusIdentifier else {
                    return
                }
                let managedLocationManager = PWManagedLocationManager(campusId: campusIdentifier)
                self?.mapView.register(managedLocationManager)
            }
        }
        else {
            DispatchQueue.main.async { [weak self] in
                guard let buildingIdentifier = self?.buildingIdentifier else {
                    return
                }
                let managedLocationManager = PWManagedLocationManager(buildingId: buildingIdentifier)
                self?.mapView.register(managedLocationManager)
            }
        }
    }
    
    func configureMapViewConstraints() {
        mapView.translatesAutoresizingMaskIntoConstraints = false
        mapView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        mapView.bottomAnchor.constraint(equalTo: toolbar.topAnchor).isActive = true
        mapView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        mapView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
    }
    
    @IBAction func trackingModeButtonTapped(_ sender: Any) {
        switch mapView.trackingMode {
        case .none:
            mapView.trackingMode = .follow
        case .follow:
            mapView.trackingMode = .followWithHeading
        case .followWithHeading:
            mapView.trackingMode = .none
        }
    }
}

// MARK: - PWMapViewDelegate
extension LocationModesViewController: PWMapViewDelegate {
    
    func mapView(_ mapView: PWMapView!, didChangeIndoorUserTrackingMode mode: PWTrackingMode) {
        switch mode {
        case .none:
            trackingModeButton.image = .emptyTrackingImage(color: .blue)
        case .follow:
            trackingModeButton.image = .filledTrackingImage(color: .blue)
        case .followWithHeading:
            trackingModeButton.image = .trackWithHeadingImage(color: .blue)
        }
    }
    
    func mapView(_ mapView: PWMapView!, didFailToLocateIndoorUserWithError error: Error!) {
        guard let error = error as NSError? else {
            return
        }
        
        let title = error.domain
        let description = error.userInfo["message"] as? String ?? "Unknown Error"
        let message = "\(description) \n Error Code: \(error.code)"
        
        showAlertForIndoorLocationFailure(withTitle: title , failureMessage: message)
    }
}

// MARK: - CLLocationManagerDelegate
extension LocationModesViewController: CLLocationManagerDelegate {
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        switch status {
        case .authorizedAlways, .authorizedWhenInUse:
            startManagedLocationManager()
        default:
            mapView.unregisterLocationManager()
            print("Not authorized to start PWLocationManager")
        }
    }
}

// MARK: UIImage Extension
extension UIImage {
    
    class func emptyTrackingImage(color: UIColor) -> UIImage {
        let rect = CGRect(x: 0, y: 0, width: 26, height: 26)
        
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0)
        
        let bezierPath = UIBezierPath()
        bezierPath.move(to: CGPoint(x: 13.53, y: 24))
        bezierPath.addLine(to: CGPoint(x: 13.44, y: 10.62))
        bezierPath.addLine(to: CGPoint(x: 0, y: 10.71))
        bezierPath.addLine(to: CGPoint(x: 24, y: 0))
        bezierPath.addLine(to: CGPoint(x: 13.53, y: 24))
        bezierPath.close()
        bezierPath.move(to: CGPoint(x: 14.31, y: 9.74))
        bezierPath.addLine(to: CGPoint(x: 14.38, y: 19.85))
        bezierPath.addLine(to: CGPoint(x: 22.29, y: 1.7))
        bezierPath.addLine(to: CGPoint(x: 4.16, y: 9.81))
        bezierPath.addLine(to: CGPoint(x: 14.31, y: 9.74))
        bezierPath.close()
        bezierPath.miterLimit = 4
        
        color.setFill()
        bezierPath.fill()
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        if let image = image {
            return image
        }
        return UIImage()
    }
    
    class func filledTrackingImage(color: UIColor) -> UIImage {
        let rect = CGRect(x: 0, y: 0, width: 26, height: 26)
        
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0)
        
        let bezierPath = UIBezierPath()
        bezierPath.move(to: CGPoint(x: 12.4, y: 22))
        bezierPath.addLine(to: CGPoint(x: 12.33, y: 9.74))
        bezierPath.addLine(to: CGPoint(x: 0, y: 9.81))
        bezierPath.addLine(to: CGPoint(x: 22, y: 0))
        bezierPath.addLine(to: CGPoint(x: 12.4, y: 22))
        bezierPath.close()
        bezierPath.miterLimit = 4
        
        color.setFill()
        bezierPath.fill()
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        if let image = image {
            return image
        }
        return UIImage()
    }
    
    class func trackWithHeadingImage(color: UIColor) -> UIImage {
        let rect = CGRect(x: 0, y: 0, width: 26, height: 26)
        
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0)
        
        let bezierPath = UIBezierPath()
        bezierPath.move(to: CGPoint(x: 6.98, y: 0))
        bezierPath.addLine(to: CGPoint(x: 8, y: 0))
        bezierPath.addLine(to: CGPoint(x: 8, y: 6.01))
        bezierPath.addLine(to: CGPoint(x: 6.98, y: 6.01))
        bezierPath.addLine(to: CGPoint(x: 6.98, y: 0))
        bezierPath.close()
        bezierPath.move(to: CGPoint(x: 7.44, y: 19.01))
        bezierPath.addLine(to: CGPoint(x: 0, y: 26))
        bezierPath.addLine(to: CGPoint(x: 7.39, y: 8.01))
        bezierPath.addLine(to: CGPoint(x: 15, y: 25.91))
        bezierPath.addLine(to: CGPoint(x: 7.44, y: 19.01))
        bezierPath.close()
        bezierPath.miterLimit = 4
        
        color.setFill()
        bezierPath.fill()
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        if let image = image {
            return image
        }
        return UIImage()
    }
}
